aws/resource_aws_greengrass_device_definition.go

```diff
@@ -0,0 +1,263 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"os"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/greengrass"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/keyvaluetags"
+)
+
+func resourceAwsGreengrassDeviceDefinition() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsGreengrassDeviceDefinitionCreate,
+		Read:   resourceAwsGreengrassDeviceDefinitionRead,
+		Update: resourceAwsGreengrassDeviceDefinitionUpdate,
+		Delete: resourceAwsGreengrassDeviceDefinitionDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"tags": tagsSchema(),
+			"latest_definition_version_arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"device_definition_version": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"arn": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"device": {
+							Type:     schema.TypeSet,
+							Optional: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"certificate_arn": {
+										Type:     schema.TypeString,
+										Required: true,
+									},
+									"id": {
+										Type:     schema.TypeString,
+										Required: true,
+									},
+									"sync_shadow": {
+										Type:     schema.TypeBool,
+										Optional: true,
+										Default:  false,
+									},
+									"thing_arn": {
+										Type:     schema.TypeString,
+										Required: true,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func createDeviceDefinitionVersion(d *schema.ResourceData, conn *greengrass.Greengrass) error {
+	var rawData map[string]interface{}
+	if v := d.Get("device_definition_version").(*schema.Set).List(); len(v) == 0 {
+		return nil
+	} else {
+		rawData = v[0].(map[string]interface{})
+	}
+
+	params := &greengrass.CreateDeviceDefinitionVersionInput{
+		DeviceDefinitionId: aws.String(d.Id()),
+	}
+
+	if v := os.Getenv("AMZN_CLIENT_TOKEN"); v != "" {
```

Comment:
This would be better as an attribute on the resource which could be passed as an explicit variable or using an environment variable. The `Sensitive` flag could be set on the attribute schema.